array_api_compat/common/_helpers.py

```diff
@@ -801,6 +802,257 @@ def size(x):
 
+def is_writeable_array(x):
+    """
+    Return False if x.__setitem__ is expected to raise; True otherwise
+    """
+    if is_numpy_array(x):
+        return x.flags.writeable
+    if is_jax_array(x):
+        return False
+    return True
+
+def _parse_copy_param(x, copy: bool | None) -> bool:
+    """Preprocess and validate a copy parameter, in line with the same
+    parameter in np.asarray(), np.astype(), etc.
+    """
+    if copy is None:
+        return not is_writeable_array(x)
+    if copy is False:
+        if not is_writeable_array(x):
+            raise ValueError("Cannot avoid modifying parameter in place")
+    elif copy is not True:
+        raise ValueError(f"Invalid value for copy: {copy!r}")
+    return copy
+
+_undef = object()
+
+class at:
+    """
+    Update operations for read-only arrays.
+
+    This implements ``jax.numpy.ndarray.at`` for all backends.
+    Writeable arrays may be updated in place; you should not rely on it.
+
+    Keyword arguments (e.g. ``indices_are_sorted``) are passed to JAX and are
+    quietly ignored for backends that don't support them.
+
+    Additionally, this introduces support for the `copy` keyword for all backends:
+
+    None
+        x *may* be modified in place if it is possible and beneficial
+        for performance. You should not use x after calling this function.
+    True
+        Ensure that the inputs are not modified. This is the default.
+    False
+        Raise ValueError if a copy cannot be avoided.
+
+    Examples
+    --------
+    Given either of these equivalent expressions::
+
+      x = at(x)[1].add(2, copy=None)
+      x = at(x, 1).add(2, copy=None)
+
+    If x is a JAX array, they are the same as::
+
+      x = x.at[1].add(x)
+
+    If x is a read-only numpy array, they are the same as::
+
+      x = x.copy()
+      x[1] += 2
+
+    Otherwise, they are the same as::
+
+      x[1] += 2
+
+    Warning
+    -------
+    When you use copy=None, you should always immediately overwrite
+    the parameter array::
+
+        x = at(x, 0).set(2, copy=None)
+
+    The anti-pattern below must be avoided, as it will result in different behaviour
+    on read-only versus writeable arrays:
+
+        x = xp.asarray([0, 0, 0])
+        y = at(x, 0).set(2, copy=None)
+        z = at(x, 1).set(3, copy=None)
+
+    In the above example, y == [2, 0, 0] and z == [0, 3, 0] when x is read-only,
+    whereas y == z == [2, 3, 0] when x is writeable!
+
+    See Also
+    --------
+    https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.ndarray.at.html
+    """
+    __slots__ = ("x", "idx")
+
+    def __init__(self, x, idx=_undef):
+        self.x = x
+        self.idx = idx
+
+    def __getitem__(self, idx):
+        """
+        Allow for the alternate syntax ``at(x)[start:stop:step]``,
+        which looks prettier than ``at(x, slice(start, stop, step))``
+        and feels more intuitive coming from the JAX documentation.
+        """
+        self.idx = idx
```

Comment:
I think this should error if `self.idx` is already set, otherwise people may expect `at(x)[0][1].set(2)` to be the same as `x[0][1] = 2`